OEL-1098: Theme the user profile page

donquixote opened the pull request:
[OEL-1098](https://citnet.tech.ec.europa.eu/CITnet/jira/browse/OEL-1098)

### Description

[Insert description here]

### Change log

- Added:
- Changed:
- Deprecated:
- Removed:
- Fixed:
- Security:

donquixote committed:
OEL-1098: Theme user 'full' view mode.

donquixote committed:
OEL-1098: Configure user profile display.

donquixote committed:
OEL-1098: Install twig_field_value.

drishu commented on modules/oe_showcase_user_profile/templates/user--full.html.twig:
```diff
@@ -0,0 +1,47 @@
+    title: content.field_first_name|field_value|render ~ ' ' ~ content.field_last_name|field_value|render,
+    meta: [
+      content.group_professional_information.field_organization|field_value,
+      content.group_professional_information.field_current_position|field_value,
```
if you agree lets use the field_value filter also in the teaser (now that we have it)

donquixote replied:
I like the idea, but I think we should do it in a separate issue.
We might have to review more things about the teaser, e.g. the usage of the search variant for the card pattern.

drishu commented on modules/oe_showcase_user_profile/oe_showcase_user_profile.module:
```diff
@@ -48,3 +54,20 @@ function oe_showcase_user_profile_user_format_name_alter(string &$name, AccountI
+ * Implements hook_preprocess_HOOK().
+ */
+function oe_showcase_user_profile_preprocess_user__full(array &$variables): void {
+  // Set an image value object for the header banner pattern.
```
no need for the comment, lets comment code only if ambiguous

drishu commented on modules/oe_showcase_user_profile/oe_showcase_user_profile.module:
```diff
@@ -48,3 +54,20 @@ function oe_showcase_user_profile_user_format_name_alter(string &$name, AccountI
+ */
+function oe_showcase_user_profile_preprocess_user__full(array &$variables): void {
+  // Set an image value object for the header banner pattern.
+  $image_element = $variables['elements']['field_profile_image'][0] ?? NULL;
```
Lets rather bail out early to not execute any additional code if not needed.
instead of checking if certain keys in the render array are present it would be useful and more elegant to use the user object that we have available in the $variables and then make use of the field API
```suggestion
  $user = $variables['user'];

  if ($user->get('field_profile_image')->isEmpty()) {
    return;
  }
```

donquixote replied:
ok, in that case we should remove the image in field ui

drishu commented on modules/oe_showcase_user_profile/oe_showcase_user_profile.module:
```diff
@@ -48,3 +54,20 @@ function oe_showcase_user_profile_user_format_name_alter(string &$name, AccountI
+  // Set an image value object for the header banner pattern.
+  $image_element = $variables['elements']['field_profile_image'][0] ?? NULL;
+  $variables['pattern_image'] = NULL;
+  if ($image_element !== NULL) {
```
no need, we check and bail out if not set earlier

drishu commented on modules/oe_showcase_user_profile/oe_showcase_user_profile.module:
```diff
@@ -48,3 +54,20 @@ function oe_showcase_user_profile_user_format_name_alter(string &$name, AccountI
+function oe_showcase_user_profile_preprocess_user__full(array &$variables): void {
+  // Set an image value object for the header banner pattern.
+  $image_element = $variables['elements']['field_profile_image'][0] ?? NULL;
+  $variables['pattern_image'] = NULL;
```
no need, if there is an image then we have will guarantee a value

drishu commented on modules/oe_showcase_user_profile/oe_showcase_user_profile.module:
```diff
@@ -48,3 +54,20 @@ function oe_showcase_user_profile_user_format_name_alter(string &$name, AccountI
+  $image_element = $variables['elements']['field_profile_image'][0] ?? NULL;
+  $variables['pattern_image'] = NULL;
+  if ($image_element !== NULL) {
+    $image_item = $image_element['#item'] ?? NULL;
```
```suggestion
    $image_item = $user->get('field_profile_image')->get(0);
```

drishu commented on modules/oe_showcase_user_profile/oe_showcase_user_profile.module:
```diff
@@ -48,3 +54,20 @@ function oe_showcase_user_profile_user_format_name_alter(string &$name, AccountI
+  $variables['pattern_image'] = NULL;
+  if ($image_element !== NULL) {
+    $image_item = $image_element['#item'] ?? NULL;
+    if ($image_item instanceof ImageItem) {
```
bail out again, this will make the code more readable
```suggestion
    if (!$image_item instanceof ImageItem) {
    return;
  }
```

drishu commented on modules/oe_showcase_user_profile/oe_showcase_user_profile.module:
```diff
@@ -48,3 +54,20 @@ function oe_showcase_user_profile_user_format_name_alter(string &$name, AccountI
+        : ImageValueObject::fromImageItem($image_item);
+    }
+  }
+}
```
since the $variables['elements']['field_profile_image'] will not be printed on page we should add the needed cache directly to variables
you can get the file entity
```
  $cacheability = CacheableMetadata::createFromRenderArray($variables['elements']['field_profile_image'][0]);
  $cacheability->applyTo($variables);
```
best after the is empty check I think

donquixote replied:
But `$variables` is not a render array.

donquixote replied:
Actually your code is correct. `$variables` is not a render array, but still this is the supported way of doing it. 

In ThemeManager::render():
```php
      foreach ($info['preprocess functions'] as $preprocessor_function) {
        if (function_exists($preprocessor_function)) {
          $preprocessor_function($variables, $hook, $info);
        }
      }
      // Allow theme preprocess functions to set $variables['#attached'] and
      // $variables['#cache'] and use them like the corresponding element
      // properties on render arrays. In Drupal 8, this is the (only) officially
      // supported method of attaching bubbleable metadata from preprocess
      // functions. Assets attached here should be associated with the template
      // that we are preprocessing variables for.
      $preprocess_bubbleable = [];
      foreach (['#attached', '#cache'] as $key) {
        if (isset($variables[$key])) {
          $preprocess_bubbleable[$key] = $variables[$key];
        }
      }
```

drishu commented on composer.json:
```diff
@@ -33,7 +35,7 @@
         "openeuropa/oe_multilingual": "^1.9",
         "openeuropa/oe_paragraphs": "dev-master",
-        "openeuropa/oe_whitelabel": "0.1.202202071145"
+        "openeuropa/oe_whitelabel": "0.1.202202080830"
```
for now do as in 1.x, we'll rebase the epic at a later point

donquixote committed:
OEL-1098: Change the way the image is preprocessed, add cache info.

donquixote committed:
OEL-1098: Use description_list pattern for user field data.

drishu commented on modules/oe_showcase_user_profile/templates/user--full.html.twig:
```diff
@@ -0,0 +1,45 @@
+      content.group_professional_information.field_current_position|field_value,
+    ],
+    content: content.field_bio|field_value,
+    attributes: create_attribute().addClass('mb-4'),
```
Normally this would go into its own block and be printed in the appropriate region right after the main tag, but since we don't invest in that in v1 lets fix the left padding by adding `ps-0`

donquixote replied:
This is a bug in bootstrap itself:
https://github.com/twbs/bootstrap/discussions/35807#discussioncomment-2134560
https://github.com/twbs/bootstrap/issues/35282
I posted about this in teams already, in front-end channel.

Fixes already exist in main branch of bootstrap:
https://github.com/twbs/bootstrap/pull/35825 and some prior commits.

We should upgrade bootstrap, or patch it. Then this will go away.
We should also remove the `<div class="row">` around the `{% include ..card.. %}` in `bcl-content-banner.html.twig`. It has no place here.

donquixote replied:
New issue: https://citnet.tech.ec.europa.eu/CITnet/jira/browse/OEL-1226
This will fix the problem.
Let's ignore this here.

drishu replied:
The way we should do it is either:

1. add a fix and a comment saying "@todo: Remove once OEL-1226 is implemented.", or
2. leave a comment saying "Issue will be fixed by OEL-1226"

In this specific content I'd go with 1.

donquixote replied:
Even with the `px-0` (which is more correct than `ps-0`), this still looks wrong on mobile.
This must be fixed in bcl.
But ok, I am adding the solution 2 for now.

donquixote replied:
Btw @drishu we can't add comments inside an array, only above it.

drishu commented on modules/oe_showcase_user_profile/templates/user--full.html.twig:
```diff
@@ -0,0 +1,45 @@
+  }) }}
+
+  {% for group_name, group_label in {
+    group_personal_information: 'Personal information',
```
reusing html blocks should be done via macros not loops. Lets do a macro of the entire section, something like:
```
{% macro description_list_section(group, content, .. %}
  <section class="my-3">
  ..
{% endmacro %}

description_list_section(group1, content,... 
description_list_section(group2, content,... 
```
but name it and define it whatever way you like, the principle is important here

donquixote replied:
Thanks for the hint! I checked the article about macros, and happily find that there is no risk of polluting a global namespace. So, I will do it.

drishu commented on composer.json:
```diff
@@ -24,6 +24,7 @@
         "drupal/purge_users": "dev-3.x",
         "drupal/search_api": "^1.21",
         "drupal/search_api_autocomplete": "^1.5",
+        "drupal/twig_field_value": "^2.0",
```
this is also a dependency in whitelabel and will stay that way, we can skip it;
Also, some block config changed in whitelabel and now the tabs are duplicated, we have to adapt, not just merge code in,.. we need to rename the local_tasks block to `block.block.oe_whitelabel_local_tasks.yml` and move it to the `config/optional` folder.

donquixote replied:
Arguments are being made that if we use a dependency directly, we should make it explicit.
There are code review tools that complain if e.g. we use a class from a package that is only implicitly required.

However, arguments can also be made for the opposite, if we assume the implicit dependency will always be there.

Personally I don't like to "assume", but I am also not always consistent about it.

So I am ok to remove it, if that avoids discussion.

donquixote replied:
Arguments are being made that if we use a dependency directly, we should make it explicit.
There are code review tools that complain if e.g. we use a class from a package that is only implicitly required.

However, arguments can also be made for the opposite, if we assume the implicit dependency will always be there.

Personally I don't like to "assume", but I am also not always consistent about it.

So I am ok to remove it, if that avoids discussion.

drishu commented on modules/oe_showcase_user_profile/oe_showcase_user_profile.module:
```diff
@@ -69,3 +76,24 @@ function oe_showcase_user_profile_form_user_form_alter(&$form, FormStateInterfac
+function oe_showcase_user_profile_preprocess_user__full(array &$variables): void {
+  /** @var \Drupal\user\UserInterface $user */
+  $user = $variables['user'];
+  $image_item = $user->get('field_profile_image')->get(0);
+  if (!$image_item instanceof ImageItem) {
+    // No profile image found.
+    return;
+  }
```
we have these modules about restricting access to profile data but this is overwriting that, lets add this check
```suggestion
  $field = $user->get('field_profile_image');
  $image_item = $field->get(0);
  if (!$image_item instanceof ImageItem || !$field->access()) {
    // No profile image found.
    return;
  }
```
Also, the anonymous user is denied access by default, so without this check `$variables['elements']['field_profile_image'][0]` will not be set and visiting a users profile with anonymous will throw an error.

donquixote replied:
Then could we check `['field_profile_image'][0]` instead of the `$field->access()` check?
If no access is given, will `['field_profile_image'][0]` be empty?

As a note, I very much dislike that we have to deal with these details in site-specific code.
There are much better ways to do this.

donquixote committed:
OEL-1098: Use macro instead of foreach in twig template.

donquixote committed:
OEL-1098: Don't explicitly require twig_field_value, this is now coming for free from oe_whitelabel.

NOTE:
It can be argued that dependencies should be explicit, if they are directly used.
In the end it should follow conventions in our team. So I am following feedback from @drishu.

donquixote committed:
OEL-1098: Add field access check in preprocess.

drishu commented on modules/oe_showcase_user_profile/templates/user--full.html.twig:
```diff
@@ -0,0 +1,44 @@
+ * Theme override for user in 'full' view mode.
+ */
+#}
+{% spaceless %}
```
use apply spaceless, see:
https://www.drupal.org/docs/develop/coding-standards/twig-coding-standards#whitespace
since we support 9.2 and above now

donquixote committed:
OEL-1098: Temporary fix for bcl-content-banner padding.

drishu reviewed (changes requested):
<img width="682" alt="Screenshot 2022-03-01 at 10 28 15" src="https://user-images.githubusercontent.com/11507647/156132938-b702206d-505a-4aa3-bcfa-72c7655d32a6.png">
we need to rename config/install/block.block.tabs.yml to config/optional/block.block.oe_whitelabel_local_tasks.yml of course using the 'oe_whitelabel_local_tasks' id so we don't have duplicate tabs;

Also we now have the contact forms enabled so
<img width="682" alt="Screenshot 2022-03-01 at 10 28 37" src="https://user-images.githubusercontent.com/11507647/156133663-c820c2d5-7303-407c-aeb1-71f08e28e34a.png">
we need to have the 'contact_default_status' disabled

donquixote committed:
OEL-1098: Rename tabs block in showcase to override the tabs block from whitelabel.

donquixote committed:
OEL-1098: Use 'apply spaceless' instead of 'spaceless' in twig template.

drishu commented on modules/oe_showcase_user_profile/config/install/core.entity_view_display.user.user.full.yml:
```diff
@@ -0,0 +1,173 @@
+    label: above
+    settings:
+      timezone_override: ''
+      format_type: html_date
```
lets use the d F Y format, custom for now until OEL-1224 is resolved

donquixote committed:
OEL-1098: Disable personal contact form.

This is done in two steps:
- Disable personal contact form by default.
- Hide personal contact form checkbox in user edit form.

There does not seem to be a cleaner way to do this:
https://stackoverflow.com/questions/44358790/how-to-disable-personal-contact-form-display-in-user-registration

donquixote commented:
> we need to rename config/install/block.block.tabs.yml to config/optional/block.block.oe_whitelabel_local_tasks.yml of course using the 'oe_whitelabel_local_tasks' id so we don't have duplicate tabs;

is `/optional/` the correct place, or rather `/overrides/`?

> Also we now have the contact forms enabled so we need to have the 'contact_default_status' disabled

drishu commented on modules/oe_showcase_user_profile/config/install/core.entity_view_display.user.user.default.yml:
```diff
@@ -49,11 +49,11 @@ content:
     settings:
       timezone_override: ''
-      format_type: html_date
+      date_format: 'd F Y'
```
why change it here ? this will have no effect on the user profile page, it should have been changed in the file I requested [core.entity_view_display.user.user.full.yml](https://github.com/openeuropa/oe_showcase/pull/58/files#diff-bafdf476599083808722b2695fc2431537893d02488893b4e8a116a7e24c7ae6)

donquixote replied:
Good catch. Sorry.

drishu commented on config/install/contact.settings.yml:
```diff
@@ -0,0 +1,5 @@
+flood:
+  limit: 5
+  interval: 3600
+user_default_enabled: false
```
leave only this one `user_default_enabled: false` we don't need to export the rest of it, leave those to default values

donquixote replied:
I had not considered this as an option. Good idea.

donquixote committed:
OEL-1098: Change date format to 'd F Y'.

donquixote committed:
OEL-1098: Fix namespace in test class.

donquixote committed:
OEL-1098: Fix date format in test.
